### PR TITLE
fix: improve performance of readStringVarFromBuff

### PR DESCRIPTION
**Pull request by geyslan**
Close: #4081 

### 1. Explain what the PR does

28ec9535a **fix: improve performance of readStringVarFromBuff**

```
Optimize readStringVarFromBuff for better performance and memory usage.

- The optimizations were tested using new benchmarks:
  - Short: 33.26% faster, 42.86% less memory usage, 1 fewer allocation.
  - Medium: 28.34% faster, 54.55% less memory usage, 1 fewer allocation.
  - Long: 26.90% faster, 73.42% less memory usage, 3 fewer allocations.
  - Long: with Low Max: 19.12% faster, 48.15% less memory usage, 1 fewer
    allocation.

The overall improvements show significant gains in both execution speed
and memory efficiency. For more check eventsreader_bench_test.go.

Changes:
- Preallocated the buffer with a reasonable initial capacity to avoid
  repeated slice resizing.
- Removed TrimLeft call since conversion logic already stops at the
  first nul byte decoded.
```

### CODE REVIEW

File: pkg/bufferdecoder/eventsreader.go [hunk 1]
```diff
@@ -373,21 +372,35 @@ func readStringFromBuff(ebpfMsgDecoder *EbpfDecoder) (string, error) {
...
 	var err error
 	var char int8
-	res := make([]byte, max)
+	var res []byte
+
+	switch {
+	case max > 0:
+		res = make([]byte, 0, max)
+	case max == 0:
+		res = make([]byte, 0)
```
Comment (geyslan) [inline pkg/bufferdecoder/eventsreader.go:381]: This change was based on the comment `max length can be passed as `max` to optimize memory allocation, otherwise pass 0`, but the logic below does the opposite when max is set as 0. The "Zero max length" unit test is the proof of it.

Then I ask what should be the expected behaviour when max is set as 0.

File: pkg/bufferdecoder/eventsreader.go [hunk 2]
```diff
@@ -373,21 +372,35 @@ func readStringFromBuff(ebpfMsgDecoder *EbpfDecoder) (string, error) {
...
 		res = append(res, byte(char))
+
+		// decode next char
 		err = decoder.DecodeInt8(&char)
 		if err != nil {
 			return "", errfmt.Errorf("error reading null terminated string: %v", err)
 		}
+		count++
 	}
-	res = bytes.TrimLeft(res[:], "\000")
```
[the remark this answers is not in the document]
Comment (geyslan) [reply pkg/bufferdecoder/eventsreader.go:389]: But this garbage would be from the previous decoded field, right? If so, wouldn't it be an error on the previous decoding, which isn't setting the cursor correctly?

Other point, if it starts with a nul byte, it ignores conversion and only updates the cursor, see: https://github.com/aquasecurity/tracee/pull/4194/files#diff-8ab83d2990ed3ba842ac33ac90111a04e6f95b0a1c98decf5565349faa17b355R233-R238.

[the remark this answers is not in the document]
Comment (NDStrahilevitz) [reply pkg/bufferdecoder/eventsreader.go:389]: No idea. I assume we are setting the cursor correctly as of now, unless this PR fixed some bug in regard to this. Therefore, its likely this is garbage data which can "naturally" (parentheses intended) occur. As far as I can tell, there was no actual change to the cursor logic (except cosmetic changes), so if this occurred before, it still should. I wouldn't just get rid of it without confirming it's not there anymore. I personally don't think its worth the effort of validation either, and its best to just leave as is.

Also if I understand you test right, if the TrimLeft would take effect then now the resulting string would be empty? Whereas before we would find a relevant string. If anything, it seems this test breaks previous behavior if I understand it correctly.

[the remark this answers is not in the document]
Comment (geyslan) [reply pkg/bufferdecoder/eventsreader.go:389]: > Also if I understand you test right, if the TrimLeft would take effect then now the resulting string would be empty? Whereas before we would find a relevant string. If anything, it seems this test breaks previous behavior if I understand it correctly.

The funny thing is that I didn't change behaviour, only optimized, that's why I removed the trim. As you can see in the old/current code, the loop check if the read char (first or not) is a nul, if so break the loop, so I assumed that trim is irrelevant.

The tests in this PR pass for the old code as well.

```go
func readStringVarFromBuff(decoder *EbpfDecoder, max int) (string, error) {
	var err error
	var char int8
	res := make([]byte, max)
	err = decoder.DecodeInt8(&char)
	if err != nil {
		return "", errfmt.Errorf("error reading null terminated string: %v", err)
	}
	var count int
	for count = 1; char != 0 && count < max; count++ {
		res = append(res, byte(char))
		err = decoder.DecodeInt8(&char)
		if err != nil {
			return "", errfmt.Errorf("error reading null terminated string: %v", err)
		}
	}
	res = bytes.TrimLeft(res[:], "\000")
	decoder.cursor += max - count // move cursor to end of buffer
	return string(res), nil
}
```

File: pkg/bufferdecoder/eventsreader.go (hunk 1 above)
Comment (geyslan) [reply pkg/bufferdecoder/eventsreader.go]: Removed validation.

File: pkg/bufferdecoder/eventsreader_test.go [hunk 3]
```diff
@@ -188,6 +188,88 @@ func TestReadArgFromBuff(t *testing.T) {
...
+			expectError:    false,
+		},
+		{
+			name:           "Zero max length",
+			buffer:         []byte{'H', 'e', 'l', 'l', 'o', 0, 'W', 'o', 'r', 'l', 'd'},
+			max:            0,
+			expected:       "",
+			expectedCursor: 0,
+			expectError:    false,
+		},
```
Comment (geyslan) [inline pkg/bufferdecoder/eventsreader_test.go:231]: If we set max 0, an empty string should be expected? Or should the helper disregard the max and check the buffer until finding a nul char?

File: pkg/bufferdecoder/eventsreader.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (NDStrahilevitz) [reply pkg/bufferdecoder/eventsreader.go:389]: Ok from looking around changes made related to the logic, it appears to be related to unix socket addresses. Could you try running this and triggering a unix socket connect event for example? If it seems ok, we can remove this, or at least comment it out explaining we're not sure what exactly the role of this trim was (seems to be there for at least 3 years from the blame).

[the remark this answers is not in the document]
Comment (geyslan) [reply pkg/bufferdecoder/eventsreader.go:389]: Going to leave the trim line commented. 👍🏼 

Tested and it seems ok:

```
sudo ./dist/tracee -e connect -s comm=nc
TIME             UID    COMM             PID     TID     RET              EVENT                     ARGS
11:11:46:890618  1000   nc               127182  127182  0                connect                   sockfd: 3, addr: map[sa_family:AF_UNIX sun_path:/tmp/unix_socket], addrlen: 110
```

```
nc -lU /tmp/unix_socket
nc -U /tmp/unix_socket
```

**Comment by rscampos** [comment]
Double check the benchmark:

> /usr/local/go/bin/go test -benchmem -run=^$ -bench ^(BenchmarkReadStringVarFromBuff_ShortString_warm|BenchmarkReadStringVarFromBuff_ShortString_old|BenchmarkReadStringVarFromBuff_ShortString|BenchmarkReadStringVarFromBuff_MediumString_old|BenchmarkReadStringVarFromBuff_MediumString|BenchmarkReadStringVarFromBuff_LongString_old|BenchmarkReadStringVarFromBuff_LongString|BenchmarkReadStringVarFromBuff_LongStringLowMax_old|BenchmarkReadStringVarFromBuff_LongStringLowMax)$ github.com/aquasecurity/tracee/pkg/bufferdecoder -benchtime=1000000x

```
=== RUN   BenchmarkReadStringVarFromBuff_ShortString_old
BenchmarkReadStringVarFromBuff_ShortString_old
BenchmarkReadStringVarFromBuff_ShortString_old-4                 1000000                52.02 ns/op           40 B/op          3 allocs/op
=== RUN   BenchmarkReadStringVarFromBuff_ShortString
BenchmarkReadStringVarFromBuff_ShortString
BenchmarkReadStringVarFromBuff_ShortString-4                     1000000                27.85 ns/op           16 B/op          2 allocs/op
=== RUN   BenchmarkReadStringVarFromBuff_MediumString_old
BenchmarkReadStringVarFromBuff_MediumString_old
BenchmarkReadStringVarFromBuff_MediumString_old-4                1000000                83.48 ns/op           88 B/op          3 allocs/op
=== RUN   BenchmarkReadStringVarFromBuff_MediumString
BenchmarkReadStringVarFromBuff_MediumString
BenchmarkReadStringVarFromBuff_MediumString-4                    1000000                49.47 ns/op           40 B/op          2 allocs/op
=== RUN   BenchmarkReadStringVarFromBuff_LongString_old
BenchmarkReadStringVarFromBuff_LongString_old
BenchmarkReadStringVarFromBuff_LongString_old-4                  1000000             24929 ns/op           77057 B/op          5 allocs/op
=== RUN   BenchmarkReadStringVarFromBuff_LongString
BenchmarkReadStringVarFromBuff_LongString
BenchmarkReadStringVarFromBuff_LongString-4                      1000000             13190 ns/op           20480 B/op          2 allocs/op
=== RUN   BenchmarkReadStringVarFromBuff_LongStringLowMax_old
BenchmarkReadStringVarFromBuff_LongStringLowMax_old
BenchmarkReadStringVarFromBuff_LongStringLowMax_old-4            1000000               289.4 ns/op           432 B/op          3 allocs/op
=== RUN   BenchmarkReadStringVarFromBuff_LongStringLowMax
BenchmarkReadStringVarFromBuff_LongStringLowMax
BenchmarkReadStringVarFromBuff_LongStringLowMax-4                1000000               182.8 ns/op           224 B/op          2 allocs/op

```
The results are aligned with yours:
```
Short: 46.46% faster, 60.00% less memory usage, 1 fewer allocation.
Medium: 40.74% faster, 54.55% less memory usage, 1 fewer allocation.
Long: 47.09% faster, 73.42% less memory usage, 3 fewer allocations.
Long with Low Max: 36.83% faster, 48.15% less memory usage, 1 fewer allocation.
```

**Review by rscampos — APPROVED** [review]
LGTM
